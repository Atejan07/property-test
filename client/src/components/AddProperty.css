#firstTitle {
  margin-top: 7rem;
}
.addProperty {
  padding: 2rem;
  background-color: var(--tertiary);
  color: white;
}

.formContainer {
  display: flex;
  flex-direction: column;
  padding: 1rem;
  margin: 1rem;
  border-radius: 8px;
}

.formContainer label {
  margin: 0.5rem 0 0 0.5rem;
  font-weight: bold;
  color: white;
}

.formContainer input,
.formContainer select,
.formContainer textarea {
  margin: 0.5rem 0.5rem 0 0.5rem;
  padding: 0.5rem;
  border-radius: 5px;
  min-width: 30vw;
  font-family: inherit;
  border: 1px solid var(--tertiary);
  background-color: var(--secondary);
}

.formContainer input[type='file'] {
  margin: 0;
  border: none;
  background-color: transparent;
}

.formContainer select {
  -webkit-appearance: none;
  appearance: none;
  font-size: 0.9rem;
  color: var(--tertiary);
}

input::file-selector-button {
  padding: 0.5rem 1rem;
  border: 1px solid var(--tertiary);
  border-radius: 5px;
  font-size: 0.9rem;
  font-family: inherit;
  color: var(--tertiary);
  background-color: var(--secondary);
}

input:focus,
textarea:focus,
select {
  color: var(--tertiary);
}
::placeholder {
  /* Chrome, Firefox, Opera, Safari 10.1+ */
  color: var(--tertiary);
  opacity: 1; /* Firefox */
}
:-ms-input-placeholder {
  /* Internet Explorer 10-11 */
  color: var(--tertiary);
}

::-ms-input-placeholder {
  /* Microsoft Edge */
  color: var(--tertiary);
}

.formFields {
  display: flex;
  justify-content: space-around;
  gap: 3rem;
}

.columnField h2 {
  margin-left: 0.5rem;
  font-weight: bold;
}

.formField {
  display: flex;
  flex-direction: column;
  justify-content: center;
}

.buttonContainer {
  display: flex;
  justify-content: center;
  margin-top: 1rem;
}

.buttonContainer button {
  margin: 1rem;
  padding: 10px 20px;
  border: none;
  border-radius: 5px;
  font-size: 1rem;
  font-weight: 100;
  letter-spacing: 1px;
  color: white;
  cursor: pointer;
  background-color: var(--primary);
  transition-duration: 0.3s;
}

button:hover {
  opacity: 0.8;
  color: var(--tertiary);
}

.errorMessage {
  margin-left: 0.5rem;
  color: var(--primary);
}

@media (max-width: 850px) {
  .addProperty {
    padding: 2rem 0;
  }

  .formContainer {
    padding: 0;
  }

  .formFields {
    flex-wrap: wrap;
    flex-direction: column;
    gap: 1rem;
    max-height: none;
  }

  .formField {
    width: 100%;
  }
}
