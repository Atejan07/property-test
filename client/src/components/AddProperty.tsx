import { ChangeEvent, FormEvent, useState } from 'react';
import { ToastContainer, toast } from 'react-toastify';
import { validateForm } from '../helperFunctions/helperFunctions';
import apiService from '../API/apiServices';
import { Error, Property, PropertyTypes, propertyTypes } from '../types/types';
import 'react-toastify/dist/ReactToastify.css';
import './AddProperty.css';

interface AddPropertyProps {
  updateProperties: (newProperty: Property) => void;
}

const initialState: Property = {
  id: '',
  houseNumber: '',
  streetName: '',
  city: '',
  country: '',
  postalCode: '',
  propertyType: 'Detached',
  bedrooms: 1,
  bathrooms: 1,
  description: '',
  photo: '',
  price: '',
};

const initialErrorState: Error = {
  houseNumber: '',
  streetName: '',
  city: '',
  country: '',
  postalCode: '',
  bedrooms: '',
  bathrooms: '',
  description: '',
  photo: '',
  price: '',
};

function AddProperty({ updateProperties }: AddPropertyProps) {
  const [newProperty, setNewProperty] = useState<Property>(initialState);
  const [propertyType, setPropertyType] = useState<PropertyTypes>('Detached');

  const [errors, setErrors] = useState(initialErrorState);

  const notifySuccess = (message: string) => toast.success(message);
  const notifyError = (message: string) => toast.error(message);

  function handleChange(
    e: ChangeEvent<HTMLInputElement | HTMLTextAreaElement>
  ) {
    const { name, value } = e.target;
    setNewProperty((prevProperties) => ({
      ...prevProperties,
      [name as string]: value,
    }));
  }

  async function handleImage(fileInput: HTMLInputElement) {
    try {
      if (!fileInput || !fileInput.files || fileInput.files.length === 0)
        return;
      const file = fileInput.files[0];
      const data = new FormData();
      data.append('my_file', file);
      const res = await apiService.saveImage(data);
      return res.url;
    } catch (error) {
      console.log(error);
    }
  }

  async function handleSubmit(
    e: FormEvent<HTMLFormElement> | ChangeEvent<HTMLInputElement>
  ) {
    try {
      e.preventDefault();

      // Validate form:
      const errors = validateForm(newProperty);
      if (Object.keys(errors).length > 0) {
        setErrors(errors);
        notifyError('Please complete all the fields correctly.');
        return;
      }

      // Handle image upload:
      const fileInput = e.currentTarget.querySelector(
        'input[type="file"]'
      ) as HTMLInputElement;
      newProperty.photo = await handleImage(fileInput);

      newProperty.propertyType = propertyType;
      const addPropertyResponse = await apiService.addProperty(newProperty);
      const { message, response } = addPropertyResponse;

      // Clear inputs:
      setNewProperty(initialState);

      if (response) {
        // Handle error:
        const errorMessage = response.data.message;
        notifyError(errorMessage);
      } else {
        notifySuccess(message);
        updateProperties(newProperty);
        setErrors(initialErrorState);
        window.scrollTo(0, 0);
      }
    } catch (error) {
      console.log(error);
    }
  }

  return (
    <div className='addProperty'>
      <ToastContainer
        position='top-left'
        autoClose={3000}
        hideProgressBar={false}
        newestOnTop={false}
        closeOnClick
        rtl={false}
        pauseOnFocusLoss
        draggable
        pauseOnHover
        theme='light'
      />
      <h1>Add your property</h1>
      <div className='formContainer'>
        <form className='formContainer' onSubmit={handleSubmit}>
          <div className='formFields'>
            <div className='columnField'>
              <div className='formField'>
                <h2>Address</h2>
                <label htmlFor='houseNumber'>House Number</label>
                <input
                  id='houseNumber'
                  type='text'
                  name='houseNumber'
                  value={newProperty.houseNumber}
                  onChange={handleChange}
                />
              </div>
              <div className='errorMessage'>{errors && errors.houseNumber}</div>
              <div className='formField'>
                <label htmlFor='streetName'>Street Name</label>
                <input
                  id='streetName'
                  type='text'
                  name='streetName'
                  value={newProperty.streetName}
                  onChange={handleChange}
                />
              </div>
              <div className='errorMessage'>{errors && errors.streetName}</div>
              <div className='formField'>
                <label htmlFor='city'>City</label>
                <input
                  id='city'
                  type='text'
                  name='city'
                  value={newProperty.city}
                  onChange={handleChange}
                />
              </div>
              <div className='errorMessage'>{errors && errors.city}</div>
              <div className='formField'>
                <label htmlFor='country'>Country</label>
                <input
                  id='country'
                  type='text'
                  name='country'
                  value={newProperty.country}
                  onChange={handleChange}
                />
              </div>
              <div className='errorMessage'>{errors && errors.country}</div>
              <div className='formField'>
                <label htmlFor='postalCode'>Postal Code</label>
                <input
                  id='postalCode'
                  type='text'
                  name='postalCode'
                  value={newProperty.postalCode}
                  onChange={handleChange}
                />
              </div>
              <div className='errorMessage'>{errors && errors.postalCode}</div>
            </div>
            <div className='columnField'>
              <h2>Details</h2>
              <div className='formField'>
                <label htmlFor='propertyType'>Property Type</label>
                <select
                  id='propertyType'
                  name='propertyType'
                  onChange={(e) =>
                    setPropertyType(e.target.value as PropertyTypes)
                  }
                >
                  {propertyTypes.map((type) => (
                    <option key={type} value={type}>
                      {type}
                    </option>
                  ))}
                </select>
              </div>
              <div className='formField'>
                <label htmlFor='bedrooms'>Bedrooms</label>
                <input
                  id='bedrooms'
                  type='number'
                  name='bedrooms'
                  value={newProperty.bedrooms}
                  onChange={handleChange}
                />
              </div>
              <div className='errorMessage'>{errors && errors.bedrooms}</div>
              <div className='formField'>
                <label htmlFor='bathrooms'>Bathrooms</label>
                <input
                  id='bathrooms'
                  type='number'
                  name='bathrooms'
                  value={newProperty.bathrooms}
                  onChange={handleChange}
                />
              </div>
              <div className='errorMessage'>{errors && errors.bathrooms}</div>
              <div className='formField'>
                <label htmlFor='description'>Description</label>
                <textarea
                  id='description'
                  name='description'
                  rows={5}
                  value={newProperty.description}
                  onChange={handleChange}
                />
              </div>
              <div className='errorMessage'>{errors && errors.description}</div>
              <div className='formField'>
                <label htmlFor='price'>Price</label>
                <input
                  id='price'
                  type='text'
                  name='price'
                  value={newProperty.price}
                  onChange={handleChange}
                />
              </div>
              <div className='errorMessage'>{errors && errors.price}</div>
              <div className='formField'>
                <label htmlFor='photo'>Photo</label>
                <input
                  id='photo'
                  type='file'
                  name='photo'
                  accept='image/*'
                  value={newProperty.photo}
                  onChange={handleChange}
                />
              </div>
              <div className='errorMessage'>{errors && errors.photo}</div>
            </div>
          </div>
          <div className='buttonContainer'>
            <button type='submit'>Add property</button>
          </div>
        </form>
      </div>
    </div>
  );
}

export default AddProperty;
