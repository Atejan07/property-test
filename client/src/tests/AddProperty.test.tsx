import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import AddProperty from '../components/AddProperty';

test('renders correctly', () => {
  const { container } = render(<AddProperty updateProperties={() => {}} />);
  expect(container).toMatchSnapshot();
});

test('has the correct heading', () => {
  render(<AddProperty updateProperties={() => {}} />);
  expect(
    screen.getByRole('heading', { name: 'Add your property' })
  ).toBeInTheDocument();
});

test('has properties labels', () => {
  render(<AddProperty updateProperties={() => {}} />);
  expect(screen.getByText('House Number')).toBeInTheDocument();
  expect(screen.getByText('Street Name')).toBeInTheDocument();
  expect(screen.getByText('City')).toBeInTheDocument();
  expect(screen.getByText('Country')).toBeInTheDocument();
  expect(screen.getByText('Postal Code')).toBeInTheDocument();
  expect(screen.getByText('Property Type')).toBeInTheDocument();
  expect(screen.getByText('Bedrooms')).toBeInTheDocument();
  expect(screen.getByText('Bathrooms')).toBeInTheDocument();
  expect(screen.getByText('Description')).toBeInTheDocument();
  expect(screen.getByText('Price')).toBeInTheDocument();
  expect(screen.getByText('Photo')).toBeInTheDocument();
});

test('has properties inputs', () => {
  render(<AddProperty updateProperties={() => {}} />);
  const inputNames = [
    'House Number',
    'Street Name',
    'City',
    'Country',
    'Postal Code',
    'Description',
    'Price',
  ];
  for (const inputName of inputNames) {
    const inputElement = screen.getByRole('textbox', { name: inputName });
    const selectElement = screen.getByRole('combobox', {
      name: 'Property Type',
    });
    const optionElement = screen.getByRole('option', { name: 'Flat' });
    const spinButtons = screen.getByRole('spinbutton', { name: 'Bedrooms' });
    expect(inputElement).toBeInTheDocument();
    expect(selectElement).toBeInTheDocument();
    expect(optionElement).toBeInTheDocument();
    expect(spinButtons).toBeInTheDocument();
  }
});

it('updates house number field correctly', () => {
  render(<AddProperty updateProperties={() => {}} />);
  const houseNumberInput = screen.getByLabelText('House Number');
  userEvent.type(houseNumberInput, '123');
  expect(houseNumberInput).toHaveValue('123');
});

it('updates description field correctly', () => {
  render(<AddProperty updateProperties={() => {}} />);
  const descriptionInput = screen.getByLabelText('Description');
  userEvent.type(descriptionInput, 'Lovely house with garden.');
  expect(descriptionInput).toHaveValue('Lovely house with garden.');
});

it('selects property type option correctly', () => {
  render(<AddProperty updateProperties={() => {}} />);
  const propertyTypeSelect = screen.getByLabelText('Property Type');
  userEvent.selectOptions(propertyTypeSelect, 'Flat');
  expect(propertyTypeSelect).toHaveValue('Flat');
});

test('has button to add property', () => {
  render(<AddProperty updateProperties={() => {}} />);
  const button = screen.getByRole('button', { name: 'Add property' });
  expect(button).toBeInTheDocument();
});

test('button add the property', () => {
  render(<AddProperty updateProperties={() => {}} />);
  userEvent.type(screen.getByLabelText('House Number'), '123');
  userEvent.type(screen.getByLabelText('Street Name'), 'Main St');
  userEvent.type(screen.getByLabelText('City'), 'Cityville');
  userEvent.type(screen.getByLabelText('Country'), 'Countryland');
  userEvent.type(screen.getByLabelText('Postal Code'), '12345');
  userEvent.selectOptions(screen.getByLabelText('Property Type'), 'Detached');
  userEvent.type(screen.getByLabelText('Bedrooms'), '3');
  userEvent.type(screen.getByLabelText('Bathrooms'), '2');
  userEvent.type(
    screen.getByLabelText('Description'),
    'This is a test description'
  );
  userEvent.type(screen.getByLabelText('Price'), '100000');
  userEvent.type(
    screen.getByLabelText('Photo'),
    'https://images.pexels.com/photos/186077/pexels-photo-186077.jpeg?auto=compress&cs=tinysrgb&w=1260&h=750&dpr=2'
  );
  userEvent.click(screen.getByText('Add property'));

  expect(screen.getByText('This is a test description')).toBeInTheDocument();
});
