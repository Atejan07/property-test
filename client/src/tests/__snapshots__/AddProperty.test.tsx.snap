// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div>
  <div
    class="addProperty"
  >
    <div
      class="Toastify"
    />
    <h1>
      Add your property
    </h1>
    <div
      class="formContainer"
    >
      <form
        class="formContainer"
      >
        <div
          class="formFields"
        >
          <div
            class="columnField"
          >
            <div
              class="formField"
            >
              <h2>
                Address
              </h2>
              <label
                for="houseNumber"
              >
                House Number
              </label>
              <input
                id="houseNumber"
                name="houseNumber"
                type="text"
                value=""
              />
            </div>
            <div
              class="errorMessage"
            />
            <div
              class="formField"
            >
              <label
                for="streetName"
              >
                Street Name
              </label>
              <input
                id="streetName"
                name="streetName"
                type="text"
                value=""
              />
            </div>
            <div
              class="errorMessage"
            />
            <div
              class="formField"
            >
              <label
                for="city"
              >
                City
              </label>
              <input
                id="city"
                name="city"
                type="text"
                value=""
              />
            </div>
            <div
              class="errorMessage"
            />
            <div
              class="formField"
            >
              <label
                for="country"
              >
                Country
              </label>
              <input
                id="country"
                name="country"
                type="text"
                value=""
              />
            </div>
            <div
              class="errorMessage"
            />
            <div
              class="formField"
            >
              <label
                for="postalCode"
              >
                Postal Code
              </label>
              <input
                id="postalCode"
                name="postalCode"
                type="text"
                value=""
              />
            </div>
            <div
              class="errorMessage"
            />
          </div>
          <div
            class="columnField"
          >
            <h2>
              Details
            </h2>
            <div
              class="formField"
            >
              <label
                for="propertyType"
              >
                Property Type
              </label>
              <select
                id="propertyType"
                name="propertyType"
              >
                <option
                  value="Detached"
                >
                  Detached
                </option>
                <option
                  value="Semi-detached"
                >
                  Semi-detached
                </option>
                <option
                  value="Terraced"
                >
                  Terraced
                </option>
                <option
                  value="Flat"
                >
                  Flat
                </option>
                <option
                  value="Bungalow"
                >
                  Bungalow
                </option>
                <option
                  value="Farm/land"
                >
                  Farm/land
                </option>
                <option
                  value="Park home"
                >
                  Park home
                </option>
              </select>
            </div>
            <div
              class="formField"
            >
              <label
                for="bedrooms"
              >
                Bedrooms
              </label>
              <input
                id="bedrooms"
                name="bedrooms"
                type="number"
                value="1"
              />
            </div>
            <div
              class="errorMessage"
            />
            <div
              class="formField"
            >
              <label
                for="bathrooms"
              >
                Bathrooms
              </label>
              <input
                id="bathrooms"
                name="bathrooms"
                type="number"
                value="1"
              />
            </div>
            <div
              class="errorMessage"
            />
            <div
              class="formField"
            >
              <label
                for="description"
              >
                Description
              </label>
              <textarea
                id="description"
                name="description"
                rows="5"
              />
            </div>
            <div
              class="errorMessage"
            />
            <div
              class="formField"
            >
              <label
                for="price"
              >
                Price
              </label>
              <input
                id="price"
                name="price"
                type="text"
                value=""
              />
            </div>
            <div
              class="errorMessage"
            />
            <div
              class="formField"
            >
              <label
                for="photo"
              >
                Photo
              </label>
              <input
                accept="image/*"
                id="photo"
                name="photo"
                type="file"
                value=""
              />
            </div>
            <div
              class="errorMessage"
            />
          </div>
        </div>
        <div
          class="buttonContainer"
        >
          <button
            type="submit"
          >
            Add property
          </button>
        </div>
      </form>
    </div>
  </div>
</div>
`;
